internal/testrunner/runners/system/servicedeployer/factory.go

```diff
@@ -5,27 +5,62 @@
 package servicedeployer
 
 import (
-	"errors"
+	"fmt"
 	"os"
 	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/elastic-package/internal/logger"
 )
 
+const devDeployDir = "_dev/deploy"
+
 var (
 	// ErrNotFound is returned when the appropriate service runner for a package
 	// cannot be found.
 	ErrNotFound = errors.New("unable to find service runner")
 )
 
-// Factory chooses the appropriate service runner for the given package, depending
-// on service configuration files defined in the package.
-func Factory(packageRootPath string) (ServiceDeployer, error) {
-	packageDevPath := filepath.Join(packageRootPath, "_dev")
+// FactoryOptions defines options used to create an instance of a service deployer.
+type FactoryOptions struct {
+	PackageRootPath    string
+	DataStreamRootPath string
+}
+
+// Factory chooses the appropriate service runner for the given data stream, depending
+// on service configuration files defined in the package or data stream.
+func Factory(options FactoryOptions) (ServiceDeployer, error) {
+	devDeployPath, err := findDevDeployPath(options)
+	if err != nil {
+		logger.Errorf("can't find _dev/deploy directory")
+		return nil, ErrNotFound
+	}
+	fmt.Println(devDeployPath)
 
 	// Is the service defined using a docker compose configuration file?
-	dockerComposeYMLPath := filepath.Join(packageDevPath, "deploy", "docker", "docker-compose.yml")
+	dockerComposeYMLPath := filepath.Join(devDeployPath, "docker", "docker-compose.yml")
 	if _, err := os.Stat(dockerComposeYMLPath); err == nil {
 		return NewDockerComposeServiceDeployer(dockerComposeYMLPath)
 	}
-
 	return nil, ErrNotFound
 }
+
+func findDevDeployPath(options FactoryOptions) (string, error) {
+	dataStreamDevDeployPath := filepath.Join(options.DataStreamRootPath, devDeployDir)
+	_, err := os.Stat(dataStreamDevDeployPath)
+	if err == nil {
+		return dataStreamDevDeployPath, nil
+	} else if !os.IsNotExist(err) {
+		return "", errors.Wrapf(err, "stat failed (path: %s)", dataStreamDevDeployPath)
+	}
+
+	packageDevDeployPath := filepath.Join(options.PackageRootPath, devDeployDir)
+	_, err = os.Stat(packageDevDeployPath)
+	if err == nil {
+		return packageDevDeployPath, nil
+	} else if !os.IsNotExist(err) {
+		return "", errors.Wrapf(err, "stat failed (path: %s)", packageDevDeployPath)
```

Comment:
Both the errors (the one and the one above for the data stream) are very similar. Can you differentiate them a bit by saying that one failed for the data stream and one for the package?

Reply:
The path argument would be different, but I extended the error message to include also some context.